File: programmingModels/OpenMP/fortran/03_map.f90
program main
  implicit none
  double precision, allocatable :: a(:)
  double precision, allocatable :: b(:)
  double precision  scalar
  integer err, i, j
  integer num_errors
  integer num_elements

  scalar = 8d0
  num_errors = 0
  num_elements = 1024

  allocate (a(num_elements),stat=err)
  allocate (b(num_elements),stat=err)

  ! initialize on the host
  do j=1,num_elements
     a(j) = 0d0
     b(j) = j
  end do


!$omp target teams distribute parallel do simd map(tofrom:a) map(to:b)
    do j=1,num_elements
       a(j) = a(j)+scalar*b(j)
    end do
!$omp end target teams distribute parallel do simd


  ! error checking
  do j=1,num_elements
     if( abs(a(j) - j*scalar) .gt. 0.000001 ) then
        num_errors = num_errors + 1
     end if

  end do

  deallocate(a)
  deallocate(b)

  if(num_errors == 0) then
    write(*,*) "Success!\n"
  else
    write(*,*) "Wrong!\n"
  endif

end program main

File: programmingModels/OpenMP/demo/cpp/03_map.cpp
#include <math.h>
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

int main( int argc, char** argv )
{

  double*   a = NULL;
  double*   b = NULL;
  double scalar = 8.0;
  int num_errors = 0;
  int num_elements = 1024;
  
  a = (double *) malloc( sizeof(double)*num_elements );
  b = (double *) malloc( sizeof(double)*num_elements );

  // initialize on the host
  for (size_t j=0; j<num_elements; j++)
    {
      a[j] = 0.0;
      b[j] = j;
    }

  //#pragma omp parallel for
  #pragma omp target teams distribute parallel for simd map(tofrom:a[:num_elements]) map(to:b[:num_elements])
  for (size_t j=0; j<num_elements; j++) {
    a[j] += scalar*b[j];
  }

  // error checking
  for (size_t j=0; j<num_elements; j++) {
    if( fabs(a[j] - (double)j*scalar) > 0.000001  ) {
      num_errors++;
    }
  }

  free(a);
  free(b);

  if(num_errors == 0) printf( "Success!\n" );

  assert(num_errors == 0);

  return 0;
}
